ArtOfIllusion/src/artofillusion/PluginRegistry.java

```diff
@@ -35,6 +35,39 @@ Some parts copyright (C) 2006 by Nik Trevallyn-Jones
   private static final HashMap<String, ExportInfo> exports = new HashMap<String, ExportInfo>();
   private static final HashMap<String, Object> classMap = new HashMap<String, Object>();
 
+  static {
+    Runtime.getRuntime().addShutdownHook(new Thread(new Runnable() {
+      @Override
+      public void run()
+      {
+        Object[] pso = new Object [0];
+        PluginRegistry.notifyPlugins(Plugin.class, Plugin.APPLICATION_STOPPING, pso);
+      }
+    }, "Plugin shutdown thread"));
+  }
+  
+  
+  public static <T> List<Throwable> notifyPlugins(Class<T> category, int message, Object... args)
+  {
+    if(categoryClasses.containsKey(category))
+    {
+        List<Throwable> errors = new ArrayList<>();
+        for(Object plugin: categoryClasses.get(category))
+        {
+          try 
+          {
+            ((Plugin)plugin).processMessage(message, args);
+          } 
+          catch(Throwable tx)
+          {
+            errors.add(tx);
+          }
+        }
+        return errors;
```

Comment:
And here, swallowing the errors without doing anything... These errors need to either be logged or bubbled up to the user.